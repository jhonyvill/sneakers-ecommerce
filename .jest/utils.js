import { render } from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import theme from "../src/Themes";

export function renderWithTheme(children) {
  return render(<ThemeProvider theme={theme}>{children}</ThemeProvider>);
}

export const productDefault = {
  id: "id-1",
  brand: "brand mock",
  title: "title mock",
  price: 250,
  discount: 0.5,
  description: "description mock",
  pictures: [
    {
      src: "image-1.jpg",
      thumb: "thumbnail-1.jpg",
      alt: "product 1",
    },
    {
      src: "image-2.jpg",
      thumb: "thumbnail-2.jpg",
      alt: "product 2",
    },
    {
      src: "image-3.jpg",
      thumb: "thumbnail-3.jpg",
      alt: "product 3",
    },
    {
      src: "image-4.jpg",
      thumb: "thumbnail-4.jpg",
      alt: "product 4",
    },
    {
      src: "image-5.jpg",
      thumb: "thumbnail-5.jpg",
      alt: "product 5",
    },
  ],
};

export const itemDefault = {
  product: { ...productDefault },
  quantity: 1,
};
