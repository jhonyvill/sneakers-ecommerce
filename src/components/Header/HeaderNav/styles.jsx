import styled from "styled-components";

export const Container = styled.nav`
  display: flex;
  align-items: center;
  gap: 1rem;
`;

export const NavLinks = styled.ul`
  display: flex;
  gap: 2rem;

`

export const StyledItem = styled.li`
  position: relative;

  &:hover a {
    color: ${({ theme }) => theme.colors.primaryOrange};
  }
  &::after {
    content: "";
    display: block;
    position: absolute;
    width: 0;
    height: 2px;
    bottom: -3rem;
    transition: 0.2s;
    background-color: ${({ theme }) => theme.colors.primaryOrange};
  }
  &:hover::after {
    width: 100%;
  }
`;