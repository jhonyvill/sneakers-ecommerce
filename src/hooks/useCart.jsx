import React from "react";
import { useState, useContext } from "react";
import { CartContext } from "../contexts/CartContext";

export const useCart = () => {
  const { cartState, cartDispatch } = useContext(CartContext);
  const [quantitySelected, setQuantitySelected] = useState(0);
  const cartHasItems = cartState.items.length > 0;

  function findProductInCart(product) {
    return cartState.items.find((item) => item.product.id === product.id);
  }

  function addCartProduct(product) {
    const existentProduct = findProductInCart(product);
    if(existentProduct){
      cartDispatch({
        type: "UPDATE_ITEM",
        payload: {
          product: product,
          quantity: (existentProduct.quantity += quantitySelected),
        },
      });
    }else if(quantitySelected > 0){
      cartDispatch({
        type: "ADD_ITEM",
        payload: { product: product, quantity: quantitySelected },
      });
    }
    setQuantitySelected(0);
  }

  return {
    quantitySelected,
    setQuantitySelected,
    addCartProduct,
    cartHasItems
  };
};
