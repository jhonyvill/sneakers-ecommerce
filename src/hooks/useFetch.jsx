import { useCallback, useState } from "react";

export const useFetch = () => {
  const [data, setData] = useState(null);
  const [error, setError] = useState(null);
  const [loading, setLoading] = useState(false);

  const request = useCallback(async (url) => {
    let response;
    let json;
    try {
      setError(null);
      setLoading(true);

      response = await fetch(url);
      json = await response.json();
      if (!response.ok) throw new Error(json.message);
    } catch (error) {
      json = null;
      setError(error.message);
      console.log(error.message);
    } finally {
      setData(json);
      setLoading(false);
    }
  }, []);

  return {
    data,
    error,
    loading,
    request,
  };
};
