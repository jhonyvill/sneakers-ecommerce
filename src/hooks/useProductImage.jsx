import React from "react";
import { useContext } from "react";
import { useState } from "react";
import { ImageProductContext } from "../contexts/ImageProductContext";

export const useProductImage = (images) => {
  const { imageSelected, indexImageSelected, handleSelectedModalImage } = useContext(ImageProductContext);
  const [featuredImage, setFeaturedImage] = useState({
    image: imageSelected || images[0],
    index: indexImageSelected,
  });

  function handleChangeImage(image, indexClicked) {
    indexClicked >= 0 &&
      indexClicked <= images.length - 1 &&
      indexClicked !== featuredImage.index &&
      setFeaturedImage({ image, index: indexClicked });
  }
  function handleClickFeaturedImage() {
    handleSelectedModalImage(featuredImage.image, featuredImage.index);
  }

  return {
    featuredImage: featuredImage.image,
    imageIndexActive: featuredImage.index,
    handleChangeImage,
    handleClickFeaturedImage,
  };
};
