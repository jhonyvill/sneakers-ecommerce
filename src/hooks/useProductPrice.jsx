import React from "react";

export const useProductPrice = (initialPrice, discount) => {
  const discountPercent = discount * 100;
  const price = initialPrice.toFixed(2);
  const discountValue = price * discount;
  const finalPrice = (price - discountValue).toFixed(2);
  const hasDiscount = discountPercent > 0;

  function getTotalItemPrice(quantity) {
    return (finalPrice * quantity).toFixed(2);
  }

  return { price, finalPrice, discountPercent, hasDiscount, getTotalItemPrice };
};
