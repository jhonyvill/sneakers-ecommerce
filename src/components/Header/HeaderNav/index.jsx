import React from "react";
import * as S from "./styles";

const HeaderNav = () => {
  return (
    <S.Container>
      <S.NavLinks>
        <S.StyledItem>
          <a href="#">Collections</a>
        </S.StyledItem>
        <S.StyledItem>
          <a href="#">Men</a>
        </S.StyledItem>
        <S.StyledItem>
          <a href="#">Women</a>
        </S.StyledItem>
        <S.StyledItem>
          <a href="#">About</a>
        </S.StyledItem>
        <S.StyledItem>
          <a href="#">Contact</a>
        </S.StyledItem>
      </S.NavLinks>
    </S.Container>
  );
};

export default HeaderNav;
