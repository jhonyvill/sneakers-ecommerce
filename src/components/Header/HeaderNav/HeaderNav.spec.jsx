import { screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { renderWithTheme } from "../../../../.jest/utils";
import * as hookMedia from "../../../hooks/useMedia";
import HeaderNav from ".";

function renderHeaderNav() {
  return renderWithTheme(<HeaderNav />);
}

function createUseMediaSpy(smallScreen) {
  return jest.spyOn(hookMedia, "useMedia").mockReturnValue(smallScreen);
}

jest.mock("../../../hooks/useMedia", () => ({
  __esModule: true,
  useMedia: () => {
    return false;
  },
}));

describe("<HeaderNav />", () => {
  it("should render HeaderNav component", () => {
    renderHeaderNav();
    
    expect(screen.getByRole('list')).toBeInTheDocument();
    expect(screen.getAllByRole('listitem')).toHaveLength(5);
    expect(screen.getByRole('link', { name: 'Men' })).toBeInTheDocument();
    expect(screen.getByRole('link', { name: 'Women' })).toBeInTheDocument();
    expect(screen.getByRole('link', { name: 'About' })).toBeInTheDocument();
    expect(screen.getByRole('link', {name: 'Contact'})).toBeInTheDocument();
    expect(screen.getByRole('link', { name: 'Collections' })).toBeInTheDocument();
    expect(screen.queryByRole('button')).not.toBeInTheDocument();
  });
  
  it("should render menu button when smallScreen is true", () => {
    createUseMediaSpy(true);
    const { getByTestId } = renderHeaderNav();

    expect(screen.getByRole('button')).toBeInTheDocument();
    expect(getByTestId("nav-list")).toHaveClass('smallScreen');
  });

  it("should change button icon when menu button is clicked", async () => {
    createUseMediaSpy(true);
    const { getByTestId } = renderHeaderNav();
    const menuButton = screen.getByRole('button');

    expect.assertions(3);
    expect(menuButton).toBeInTheDocument();
    expect(getByTestId("menu-svg")).toBeInTheDocument();

    await userEvent.click(menuButton);

    expect(getByTestId("close-svg")).toBeInTheDocument();
  });
  
  it("should change menuActive when menu button clicked", async () => {
    createUseMediaSpy(true);
    const { getByTestId } = renderHeaderNav();
    const menuButton = screen.getByRole('button');
    const navList = getByTestId("nav-list");
    await userEvent.click(menuButton);

    expect.assertions(3);
    expect(menuButton).toBeInTheDocument();
    expect(navList).toHaveClass('smallScreen');
    expect(navList).toHaveClass('menuActive'); 
  });
});

describe("<HeaderNav /> snapshots", () => {
  it("should match snapshot", () => {
    createUseMediaSpy(false);
    const { container } = renderHeaderNav();
    expect(container.firstChild).toMatchSnapshot();
  });
  it("should match snapshot small screen", () => {
    createUseMediaSpy(true);
    const { container } = renderHeaderNav();
    expect(container.firstChild).toMatchSnapshot();
  });
  
  it("should match snapshot when not small screen", () => {
    createUseMediaSpy(false);
    const { container } = renderHeaderNav();
    expect(container.firstChild).toMatchSnapshot();
  });
  it("should match snapshot small screen and menu active", async () => {
    createUseMediaSpy(true);
    const { container } = renderHeaderNav();

    const menuButton = screen.getByRole('button');
    await userEvent.click(menuButton);

    expect(container.firstChild).toMatchSnapshot();
  });
});
