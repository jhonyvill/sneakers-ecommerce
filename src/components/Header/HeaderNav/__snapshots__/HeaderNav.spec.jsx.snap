// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HeaderNav /> snapshots should match snapshot 1`] = `
<nav
  class="sc-bdnyFh yqnvy"
>
  <ul
    class="sc-gtsqUy exjatf false false"
    data-testid="nav-list"
  >
    <li
      class="sc-dlniIP bfDzoH"
    >
      <a
        href="#"
      >
        Collections
      </a>
    </li>
    <li
      class="sc-dlniIP bfDzoH"
    >
      <a
        href="#"
      >
        Men
      </a>
    </li>
    <li
      class="sc-dlniIP bfDzoH"
    >
      <a
        href="#"
      >
        Women
      </a>
    </li>
    <li
      class="sc-dlniIP bfDzoH"
    >
      <a
        href="#"
      >
        About
      </a>
    </li>
    <li
      class="sc-dlniIP bfDzoH"
    >
      <a
        href="#"
      >
        Contact
      </a>
    </li>
  </ul>
</nav>
`;

exports[`<HeaderNav /> snapshots should match snapshot small screen 1`] = `
<nav
  class="sc-bdnyFh yqnvy"
>
  <button
    aria-expanded="false"
    aria-label="menu"
    class="sc-hKFymg iXpSSQ"
  >
    <icon-mock
      data-testid="menu-svg"
    />
  </button>
  <ul
    class="sc-gtsqUy exjatf smallScreen false"
    data-testid="nav-list"
  >
    <li
      class="sc-dlniIP bfDzoH"
    >
      <a
        href="#"
      >
        Collections
      </a>
    </li>
    <li
      class="sc-dlniIP bfDzoH"
    >
      <a
        href="#"
      >
        Men
      </a>
    </li>
    <li
      class="sc-dlniIP bfDzoH"
    >
      <a
        href="#"
      >
        Women
      </a>
    </li>
    <li
      class="sc-dlniIP bfDzoH"
    >
      <a
        href="#"
      >
        About
      </a>
    </li>
    <li
      class="sc-dlniIP bfDzoH"
    >
      <a
        href="#"
      >
        Contact
      </a>
    </li>
  </ul>
</nav>
`;

exports[`<HeaderNav /> snapshots should match snapshot small screen and menu active 1`] = `
<nav
  class="sc-bdnyFh yqnvy"
>
  <button
    aria-expanded="true"
    aria-label="menu"
    class="sc-hKFymg iXpSSQ"
  >
    <icon-mock
      data-testid="close-svg"
    />
  </button>
  <ul
    class="sc-gtsqUy exjatf smallScreen menuActive"
    data-testid="nav-list"
  >
    <li
      class="sc-dlniIP bfDzoH"
    >
      <a
        href="#"
      >
        Collections
      </a>
    </li>
    <li
      class="sc-dlniIP bfDzoH"
    >
      <a
        href="#"
      >
        Men
      </a>
    </li>
    <li
      class="sc-dlniIP bfDzoH"
    >
      <a
        href="#"
      >
        Women
      </a>
    </li>
    <li
      class="sc-dlniIP bfDzoH"
    >
      <a
        href="#"
      >
        About
      </a>
    </li>
    <li
      class="sc-dlniIP bfDzoH"
    >
      <a
        href="#"
      >
        Contact
      </a>
    </li>
  </ul>
</nav>
`;

exports[`<HeaderNav /> snapshots should match snapshot when not small screen 1`] = `
<nav
  class="sc-bdnyFh yqnvy"
>
  <ul
    class="sc-gtsqUy exjatf false false"
    data-testid="nav-list"
  >
    <li
      class="sc-dlniIP bfDzoH"
    >
      <a
        href="#"
      >
        Collections
      </a>
    </li>
    <li
      class="sc-dlniIP bfDzoH"
    >
      <a
        href="#"
      >
        Men
      </a>
    </li>
    <li
      class="sc-dlniIP bfDzoH"
    >
      <a
        href="#"
      >
        Women
      </a>
    </li>
    <li
      class="sc-dlniIP bfDzoH"
    >
      <a
        href="#"
      >
        About
      </a>
    </li>
    <li
      class="sc-dlniIP bfDzoH"
    >
      <a
        href="#"
      >
        Contact
      </a>
    </li>
  </ul>
</nav>
`;
